presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java

```diff
@@ -497,8 +516,19 @@ private Optional<StatisticsFile> getClosestStatisticsFileForSnapshot(IcebergTabl
     /**
      * Builds a map of field ID to ColumnStatistics for a particular {@link StatisticsFile}.
      */
-    private Map<Integer, ColumnStatistics.Builder> loadStatisticsFile(StatisticsFile file)
+    private Map<Integer, ColumnStatistics> loadStatisticsFile(StatisticsFile file, StatisticsFileCache statisticsFileCache, List<Integer> columnIds)
     {
+        // first, try to load all stats from the cache. If the map doesn't contain all keys, load the missing
```

Comment:
Do you have an estimate of how big the puffin file gets for an Iceberg table (with say, 10s of columns) ?  

I ask because, if the file size is small enough (and given the fact that Puffin supports [compression](https://iceberg.apache.org/puffin-spec/#compression-codecs) the number of columns may not be an issue), we may be better off caching the entire puffin file in memory instead of caching at a per-column level (similar to how we cache parquet footers)

With the current implementation, repeated file IO will be performed every time we want to access a new column's stats.
This will be avoided if cache the entire `StatisticsFile`'s byte-stream instead

Reply:
> Do you have an estimate of how big the puffin file gets for an Iceberg table (with say, 10s of columns) ?

It depends on histogram configuration, but using TPC-DS SF1k with the histogram K parameter configured to 32768, each histogram consumes ~1MiB of space. So for a table with up to 100 columns which all support histograms, you'd use 100MB. The smallest configuration takes ~1KiB. If you have a large schema with many tables, it's conceivable that you could easily reach the cache limits if you stored all the columns. I think it makes sense to cache the used columns.

Also, I don't have data on the Theta sketches which we use to store NDV counts, but I believe the sketch can also consumes in the range of ~KiBs but I haven't done any thorough experimentation.

I don't have data on compression of the Puffin files. We don't currently compress the files in our writer. I think adding compression is trivial though. From an information-theoretic perspective, sketches are supposed to be information-dense which would lead me to believe that compression won't affect the size much. I need to do some experimentation though. I will update later with some results.

The difference between puffin and the parquet footers is going to be that the amount of data stored per-column is potentially much larger. I felt re-reading the puffin files for new columns was a reasonable tradeoff in order to not pollute the cache with unused data which is especially important for wide tables.

Reply:
I'm OK with starting with this fine grained cache and measuring usage. 

Specifically can we track a metric with  - 
1. Metric Value :  Size of puffin file fetched
2. Dimension : Number of columns with stats in the file
3. Dimension : File name/ File name MD5
4. Dimension : Was an existing column's stats cached aka have we read this file before (This one may be hard to track given the cache key choice)

This metric, along with cache hit/miss counts should give us a good picture of if we need to tweak what we cache

Reply:
I'm somewhat familiar with our JMX interfaces, but I'm not sure it's capable of tracking these metrics along all the  dimensions. I can record separate independent distributions for the puffin file sizes, columns amount, etc. I don't know if we have the ability to collect these so we can correlate them. Do you know of any existing examples I could look at which does something similar?

Reply:
You can log a `RuntimeMetric` using `ConnectorSession#getRuntimeStats()`, [an example](https://github.com/prestodb/presto/blob/06edc73cca2f83916e8a18d371bba56e6b32f93f/presto-hive/src/main/java/com/facebook/presto/hive/statistics/ParquetQuickStatsBuilder.java#L364-L365).  

Since we don't have a way to track dimensions on the metric, we could model the dimensions in the metric name, e.g
```
session.getRuntimeStats().addMetricValue("<TableOrPartitionIdentifier>/PuffinFileSize", RuntimeUnit.BYTE, 100_000L);
session.getRuntimeStats().addMetricValue("<TableOrPartitionIdentifier>/<FileNameOrMD5>/ColumnCount", RuntimeUnit.NONE, 10L);
session.getRuntimeStats().addMetricValue("<TableOrPartitionIdentifier>/CacheMissButOtherColumnStatsPresent", RuntimeUnit.NONE, 1L);
```

Reply:
Thanks for the pointers. I added metrics for file size and column count, as well as partial cache misses.